File: src/0713-Subarray-Product-Less-Than-K/subarray_product_less_than_k.py
from typing import List

class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0
        
        product: int = 1
        ans: int = 0
        left: int = 0

        for right, num in enumerate(nums):
            product *= num

            while product >= k:
                product /= nums[left]
                left += 1
            
            ans += right - left + 1
        
        return ans

if __name__ == "__main__":
    nums: List[int] = [10,5,2,6]
    k: int = 100
    solution = Solution()
    
    print(solution.numSubarrayProductLessThanK(nums, k))

File: src/0713-Subarray-Product-Less-Than-K/subarray_product_less_than_k.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int numSubarrayProductLessThanK(std::vector<int>& nums, int k) {
        if (k <= 1) {
            return 0;
        }
        
        int product = 1;
        int ans = 0;
        int left = 0;

        for (int right = 0; right < nums.size(); right++) {
            product *= nums[right];

            while (product >= k) {
                product /= nums[left];
                left += 1;
            }
            
            ans += right - left + 1;
        }
        
        return ans;
    }
};

int main() {
    std::vector<int> nums = {10,5,2,6};
    int k = 100;
    Solution solution;
    
    std::cout << solution.numSubarrayProductLessThanK(nums, k);

    return 0;
}